app/lib/package/export_api_to_bucket.dart

```diff
@@ -0,0 +1,297 @@
+      if (await _isSameContent(objectName, gzippedBytes)) {
+        continue;
+      }
+
+      await uploadWithRetry(
+        _bucket,
+        objectName,
+        gzippedBytes.length,
+        () => Stream.value(gzippedBytes),
+        metadata: ObjectMetadata(
+          contentType: 'application/json; charset="utf-8"',
+          contentEncoding: 'gzip',
+          cacheControl: 'public, max-age=${_pkgApiMaxCacheAge.inSeconds}',
+        ),
+      );
+    }
+  }
+
+  Future<void> _deletePackageFromBucket(String package) async {
+    for (final objectName in _apiPkgObjectNames(package)) {
+      await _bucket.tryDelete(objectName);
+    }
+  }
+
+  Stream<_PkgUpdatedEvent> _queryRecentPkgUpdatedEvents(DateTime since) async* {
+    final q1 = dbService.query<ModeratedPackage>()
+      ..filter('moderated >=', since)
+      ..order('-moderated');
+    yield* q1.run().map((mp) => mp.asPkgUpdatedEvent());
+
+    final q2 = dbService.query<Package>()
+      ..filter('updated >=', since)
+      ..order('-updated');
+    yield* q2.run().map((p) => p.asPkgUpdatedEvent());
+  }
+
+  /// Deletes obsolete runtime-versions from the bucket.
+  Future<void> deleteObsoleteRuntimeContent() async {
+    final versions = <String>{};
+
+    // Objects in the bucket are stored under the following pattern:
+    //   `current/api/<package>`
+    //   `<runtimeVersion>/api/<package>`
+    // Thus, we list with `/` as delimiter and get a list of runtimeVersions
+    await for (final d in _bucket.list(prefix: '', delimiter: '/')) {
+      if (!d.isDirectory) {
+        _logger.warning(
+            'Bucket `${_bucket.bucketName}` should not contain any top-level object: `${d.name}`');
+        continue;
+      }
+
+      // Remove trailing slash from object prefix, to get a runtimeVersion
+      if (!d.name.endsWith('/')) {
+        _logger.warning(
+            'Unexpected top-level directory name in bucket `${_bucket.bucketName}`: `${d.name}`');
+        return;
+      }
+      final rtVersion = d.name.substring(0, d.name.length - 1);
+      if (runtimeVersionPattern.matchAsPrefix(rtVersion) == null) {
+        continue;
+      }
+
+      // Check if the runtimeVersion should be GC'ed
+      if (shouldGCVersion(rtVersion)) {
+        versions.add(rtVersion);
+      }
+    }
+
+    for (final v in versions) {
+      await deleteBucketFolderRecursively(_bucket, '$v/', concurrency: 4);
+    }
+  }
+
+  Future<bool> _isSameContent(String objectName, List<int> bytes) async {
+    final info = await _bucket.tryInfo(objectName);
+    if (info == null) {
+      return false;
+    }
+    if (info.length != bytes.length) {
+      return false;
+    }
+    final md5Hash = md5.convert(bytes).bytes;
+    if (md5Hash.length != info.md5Hash.length) {
+      return false;
+    }
+    for (var i = 0; i < md5Hash.length; i++) {
+      if (md5Hash[i] != info.md5Hash[i]) {
+        return false;
+      }
+    }
+    return true;
```

Comment:
I'd suggest moving the comparison logic for comparing two byte strings that contain a hash into a `fixedTimeComparison` function -- or even better, used a fixed-time comparison function from somewhere else.

Reply:
Changed to fixed time. Although, as this is called in a background process (or maybe later *after* an upload was successful, asynchronously), I don't see why timing has any relevance here.

Reply:
Fixed time is not relevant, it's just always seems wrong to compare hashes in non-fixed time :D